peps/pep-0788.rst

```diff
@@ -224,254 +277,295 @@ created by the authors of this PEP:
 
-.. c:function:: void PyInterpreterState_Release(PyInterpreterState *interp)
+    The caller does not need to hold an :term:`attached thread state`.
 
-    Decrement the reference count of the interpreter, as was incremented by
-    :c:func:`PyInterpreterState_Hold` or :c:func:`PyInterpreterState_Lookup`.
+.. c:function:: void PyInterpreterWeakRef_Close(PyInterpreterWeakRef wref)
 
-    This function cannot fail, other than with a fatal error. The caller does
-    not need to have an :term:`attached thread state` for *interp*.
+    Release a weak reference, possibly deallocating it.
 
-Ensuring and releasing thread states
+    This function cannot fail, and the caller doesn't need to hold an
+    :term:`attached thread state`.
+
+Ensuring and Releasing Thread States
 ------------------------------------
 
 This proposal includes two new high-level threading APIs that intend to
 replace :c:func:`PyGILState_Ensure` and :c:func:`PyGILState_Release`.
 
-.. c:function:: int PyThreadState_Ensure(PyInterpreterState *interp)
-
-    Ensure that the thread has an :term:`attached thread state` for *interp*,
-    and thus can safely invoke that interpreter. It is OK to call this
-    function if the thread already has an attached thread state, as long as
-    there is a subsequent call to :c:func:`PyThreadState_Release` that matches
-    this one.
+.. c:function:: int PyThreadState_Ensure(PyInterpreterRef ref)
 
-    The reference to the interpreter *interp* is stolen by this function.
-    As such, *interp* should have been acquired by
-    :c:func:`PyInterpreterState_Hold`.
+    Ensure that the thread has an :term:`attached thread state` for the
+    interpreter denoted by *ref*, and thus can safely invoke that
+    interpreter. It is OK to call this function if the thread already has an
+    attached thread state, as long as there is a subsequent call to
+    :c:func:`PyThreadState_Release` that matches this one.
 
-    Thread states created by this function are non-daemon by default. See
-    :c:func:`PyThreadState_SetDaemon`. If the calling thread already has an
-    attached thread state that matches *interp*, then this function
-    will mark the existing thread state as non-daemon and return. It will
-    be restored to its prior daemon status upon the next
-    :c:func:`PyThreadState_Release` call.
+    Nested calls to this function will only sometimes create a new
+    :term:`thread state`. If there is no attached thread state,
+    then this function will check for the most recent attached thread
+    state used by this thread. If none exists or it doesn't match *ref*,
+    a new thread state is created. If it does match *ref*, it is reattached.
+    If there is an attached thread state, then a similar check occurs;
+    if the interpreter matches *ref*, it is attached, and otherwise a new
+    thread state is created.
 
     Return zero on success, and non-zero with the old attached thread state
```

Comment:
I expected: "Return -1 on error". I don't understand what "non-zero with the old attached thread state restored" means.